hackathon_site/hardware/serializers.py

```diff
@@ -46,9 +50,91 @@ def get_unique_hardware_count(obj: Category):
         return Hardware.objects.filter(categories__id=obj.id).count()
 
 
-class OrderSerializer(serializers.ModelSerializer):
+class OrderListSerializer(serializers.ModelSerializer):
     hardware_set = HardwareSerializer(many=True, read_only=True)
 
     class Meta:
         model = Order
         fields = ("id", "hardware_set", "team", "status", "created_at", "updated_at")
+
+
+class OrderCreateSerializer(serializers.Serializer):
+    class OrderCreateHardwareSerializer(serializers.Serializer):
+        id = serializers.PrimaryKeyRelatedField(
+            queryset=Hardware.objects.all(), many=False, required=True
+        )
+        quantity = serializers.IntegerField(required=True)
+
+    team_id = serializers.PrimaryKeyRelatedField(
+        queryset=TeamEvent.objects.all(), many=False, required=True
+    )
+    hardware = OrderCreateHardwareSerializer(many=True, required=True)
+
+    @staticmethod
+    def merge_requests(hardware_requests):
+        return functools.reduce(
+            lambda x, y: x + y,
+            [Counter({e["id"]: e["quantity"]}) for e in hardware_requests],
+        )
+
+    def create(self, validated_data):
```

Comment:
I think you can also do a `validate` method, which can creates `validated_data`. Might be a bit cleaner to move some of this logic out of there.

Thinking a bit more about that, throwing a validation error in the validate method will return a 400 from the API. We decided if there isn't enough hardware remaining we want to go ahead and place the order anyway, but how about these cases:
- Constraints weren't met
- Hardware ID doesn't exist

Should we actually abort the order for those? If so, that can go in the `validate` method. If not, disregard me .

Reply:
I think we should def catch hardware doesn't exist and throw, cuz that's just a bad request from the client. 

I'm not sure what "constraints" you are referring to? If it's the allowed quantities constraints, the design was to create order for what we can and tell the user whatever we can't fulfill. So I don't think we can just throw there

Reply:
We have two types of constraints:
- Each hardware has a `max_per_team` field. So check how many of that hardware the team has out (order items excluding orders that have been cancelled, and where `part_returned_health` is not null indicating it was returned). If how many they requested would put them over the limit, throw a 400 (`ValidationError`) with a suitable error message we can display
- Same with category, there's a `max_per_team` field. The query is probably a bit more annoying to craft, but same logic as above.

The only time we should fulfill a partial order is when some of the hardware got checked out and there isn't enough quantity remaining, because that isn't the user's fault. Them requesting more than they should be allowed is their fault (but could happen eg due to two team members submitted orders one after another without refreshing their pages).

Reply:
Added object level validation in serializer